src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.AsymmetricSignatureFormatterCmsSignature.cs

```diff
@@ -0,0 +1,70 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Pkcs.Asn1;
+using System.Security.Cryptography.X509Certificates;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    internal partial class CmsSignature
+    {
+        private sealed class AsymmetricSignatureFormatterCmsSignature : CmsSignature
+        {
+            AsymmetricSignatureFormatter _formatter;
+
+            public AsymmetricSignatureFormatterCmsSignature(AsymmetricSignatureFormatter formatter)
+            {
+                _formatter = formatter;
+            }
+
+            internal override bool VerifySignature(
+#if netcoreapp
+                ReadOnlySpan<byte> valueHash,
+                ReadOnlyMemory<byte> signature,
+#else
+                byte[] valueHash,
+                byte[] signature,
+#endif
+                string digestAlgorithmOid,
+                HashAlgorithmName digestAlgorithmName,
+                ReadOnlyMemory<byte>? signatureParameters,
+                X509Certificate2 certificate)
+            {
+                throw new NotImplementedException("This code path should never be called");
+            }
+
+            protected override bool Sign(
+#if netcoreapp
+                ReadOnlySpan<byte> dataHash,
+#else
+                byte[] dataHash,
+#endif
+                HashAlgorithmName hashAlgorithmName,
+                X509Certificate2 certificate,
+                bool silent,
+                out Oid signatureAlgorithm,
+                out byte[] signatureValue)
+            {
+                _formatter.SetHashAlgorithm(hashAlgorithmName.Name);
+
+                // Passing public key, AsymmetricAlgorithmFormatter should resolve private key on their own
+                _formatter.SetKey(certificate.PublicKey.Key);
+                signatureAlgorithm = certificate.PublicKey.Oid;
```

Comment:
This doesn't work for DSA or ECDSA.  And is not a sustainable pattern.

Reply:
@bartonjs, would `certificate.GetKeyAlgorithm()` work correctly? similarly what we use in CmsSignature.Sign for figuring out the CmsSignature instance?

Reply:
https://github.com/dotnet/corefx/blob/175283b698afb0efafe16ccd397fc4e0afa2757d/src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.DSA.cs#L101-L115

It's algorithm-specific.  AsymmetricSignatureFormatter doesn't have enough data to know how to build the identifiers that CMS expects.